apps/beeai-server/src/beeai_server/api/auth.py

```diff
@@ -1,35 +1,80 @@
 
 import jwt
-from kink import inject
+from pydantic import AwareDatetime, BaseModel
 
 from beeai_server.configuration import Configuration
-from beeai_server.domain.models.user import User
-from beeai_server.service_layer.services.users import UserService
+from beeai_server.domain.models.permissions import Permissions
+from beeai_server.domain.models.user import UserRole
 from beeai_server.utils.utils import utc_now
 
+ROLE_PERMISSIONS: dict[UserRole, Permissions] = {
+    UserRole.admin: Permissions.all(),
+    UserRole.developer: Permissions(
+        files={"*"},
+        vector_stores={"*"},
+        llm={"*"},
+        feedback={"write"},
+        embeddings={"*"},
+        a2a_proxy={"*"},
+        providers={"read", "write"},  # TODO provider ownership
+    ),
+    UserRole.user: Permissions(
+        files={"*"},
+        vector_stores={"*"},
+        llm={"*"},
+        embeddings={"*"},
+        a2a_proxy={"*"},
+        feedback={"write"},
+        providers={"read"},
+    ),
+}
+
+
+class ParsedToken(BaseModel):
+    global_permissions: Permissions
+    context_permissions: Permissions
+    context_id: UUID
+    user_id: UUID
+    raw: dict[str, Any]
 
-@inject
-def issue_token(user: User, configuration: Configuration) -> str:
-    # TODO: using admin password as secret_key - should we use a dedicated secret?
-    secret_key = configuration.auth.admin_password.get_secret_value()
+
+def issue_internal_jwt(
+    user_id: UUID,
+    context_id: UUID,
+    global_permissions: Permissions,
+    context_permissions: Permissions,
+    configuration: Configuration,
+) -> tuple[str, AwareDatetime]:
+    assert configuration.auth.jwt_secret_key
+    secret_key = configuration.auth.jwt_secret_key.get_secret_value()
     now = utc_now()
-    payload = {"user_id": str(user.id), "role": "user", "exp": now + timedelta(hours=1), "iat": now}
-    return jwt.encode(payload, secret_key, algorithm="HS256")
-
-
-@inject
-async def verify_token(token: str, configuration: Configuration, user_service: UserService) -> User:
-    # TODO: using admin password as secret_key - should we use a dedicated secret?
-    secret_key = configuration.auth.admin_password.get_secret_value()
-    try:
-        payload = jwt.decode(token, secret_key, algorithms=["HS256"])
-        return await user_service.get_user(user_id=UUID(payload["user_id"]))
-    except jwt.ExpiredSignatureError as e:
-        raise Exception("Token expired") from e
-    except jwt.InvalidTokenError as e:
-        raise Exception("Invalid token") from e
+    expires_at = now + timedelta(minutes=20)
+    payload = {
+        "user_id": str(user_id),
+        "context_id": str(context_id),
+        "exp": expires_at,
+        "iat": now,
+        "permissions": {
+            "global": global_permissions.model_dump(mode="json"),
+            "context": context_permissions.model_dump(mode="json"),
+        },
+    }
+    return jwt.encode(payload, secret_key, algorithm="HS256"), expires_at
```

Comment:
If you expect the token holder to be able to use the information from the JWT, it is probably better to use a public/private keypair for signing. The holder needs to make sure the token is valid before using that information somehow.

Reply:
right now we don't need this feature, we can change it in the future I think, this does not need a big migration :)